src/base/SanityChecker.ml

```diff
@@ -875,6 +879,165 @@ struct
+      |> fun (_impure_procedures, pure_procedures) -> pure_procedures
+
+    (** Collects a map of direct calls for procedures (
+        [procedure_name |-> set_of_callers]) and a map of calls inside each
+        component ([caller |-> set_of_calls]). *)
+    let collect_calls cg =
+      let procedures, calls =
+        CG.fold_over_nodes_dfs cg ~init:([], emp_ids_map)
+          ~f:(fun (procedures, m) node ->
+            let save_callees m =
+              let data =
+                CG.Node.succs node
+                |> List.fold_left ~init:emp_ids_set ~f:(fun s succ_node ->
+                       CG.Node.id succ_node |> get_id |> SCIdentifierSet.add s)
+              in
+              Map.set m ~data ~key:(node.id |> get_id)
+            in
+            match node.ty with
+            | Proc -> (procedures @ [ get_id node.id ], save_callees m)
+            | Trans -> (procedures, save_callees m)
+            | TFun | TFunAlias | Fun | FunAlias -> (procedures, m))
+      in
+      let procedure_callers =
+        List.fold_left procedures ~init:emp_ids_map ~f:(fun m proc_name ->
+            let data =
+              Map.fold calls ~init:emp_ids_set
+                ~f:(fun ~key:callee_name ~data:calls s ->
+                  if Set.mem calls proc_name then
+                    SCIdentifierSet.add s callee_name
+                  else s)
+            in
+            Map.set m ~key:proc_name ~data)
+      in
+      (procedure_callers, calls)
+
+    (** Reports if a component calls a pure procedure which has already been
+        called in all of its callers. *)
+    let report_redundant_calls cmod callers calls pure_procedures =
+      let is_pure id =
+        SCIdentifier.get_id id |> SCIdentifierSet.mem pure_procedures
+      in
+      let rec report_in_stmt comp_id (s, _ann) =
+        match s with
+        | MatchStmt (_id, arms) ->
+            List.iter arms ~f:(fun (_pattern, stmts) ->
+                List.iter stmts ~f:(fun sa -> report_in_stmt comp_id sa))
+        | CallProc (id, _params) when is_pure id ->
+            let pure_name = SCIdentifier.get_id id in
+            let comp_name = get_id comp_id in
+            Map.find callers comp_name
+            |> Option.value_map ~default:() ~f:(fun caller_names ->
+                   let calls_pure_name caller =
+                     Map.find calls caller
+                     |> Option.value_map ~default:false ~f:(fun call_names ->
+                            SCIdentifierSet.mem call_names pure_name)
+                   in
+                   if
+                     (not @@ Set.is_empty caller_names)
+                     && Set.for_all caller_names ~f:calls_pure_name
+                   then
+                     warn1
+                       (Printf.sprintf
+                          "Procedure without side effects %s has already been \
+                           called in all of the callers of %s. This call can \
+                           be safely removed."
```

Comment:
```suggestion
                          "Redundant call to procedure %s. This call can \
                           safely be removed."
```
Throwing an exception is technically a side-effect, so I think it's better to just say the call can be removed, without necessarily going into any details why.

Reply:
Fixed.